extmod/moddeflate.c

```diff
@@ -0,0 +1,411 @@
+        uint8_t buf[] = { 0x1f, 0x8b, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x03 };     // MTIME=0, XFL=4 (fastest), OS=3 (unix)
+        ret = stream->write(self->stream, buf, sizeof(buf), &err);
+
+        self->write->input_checksum = ~0; // CRC32
+    }
+    if (ret == MP_STREAM_ERROR) {
+        return false;
+    }
+
+    // Write starting block.
+    uzlib_start_block(&self->write->lz77);
+
+    return true;
+}
+#endif
+
+STATIC mp_obj_t deflateio_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum {
+        ARG_stream,
+        ARG_format,
+        ARG_wbits,
+        ARG_close,
+    };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_stream, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_format, MP_ARG_INT, {.u_int = DEFLATEIO_FORMAT_NONE} },
+        { MP_QSTR_wbits, MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_close, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_FALSE} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    if (args[ARG_format].u_int < DEFLATEIO_FORMAT_MIN || args[ARG_format].u_int > DEFLATEIO_FORMAT_MAX) {
+        mp_raise_ValueError(MP_ERROR_TEXT("format"));
+    }
+    if (args[ARG_wbits].u_int != 0 && (args[ARG_wbits].u_int < 5 || args[ARG_wbits].u_int > 15)) {
+        mp_raise_ValueError(MP_ERROR_TEXT("wbits"));
+    }
+
+    mp_obj_deflateio_t *self = mp_obj_malloc(mp_obj_deflateio_t, type);
+    self->stream = args[ARG_stream].u_obj;
+    self->format = args[ARG_format].u_int;
+    self->window_bits = args[ARG_wbits].u_int;
+    self->read = NULL;
+    #if MICROPY_PY_DEFLATE_COMPRESS
+    self->write = NULL;
+    #endif
+    self->close = mp_obj_is_true(args[ARG_close].u_obj);
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+STATIC mp_uint_t deflateio_read(mp_obj_t o_in, void *buf, mp_uint_t size, int *errcode) {
+    mp_obj_deflateio_t *self = MP_OBJ_TO_PTR(o_in);
+
+    if (!deflateio_init_read(self) || self->stream == MP_OBJ_NULL) {
+        *errcode = MP_EINVAL;
+        return MP_STREAM_ERROR;
+    }
+
+    if (self->read->eof) {
+        return 0;
+    }
+
+    self->read->decomp.dest = buf;
+    self->read->decomp.dest_limit = (uint8_t *)buf + size;
+    int st = uzlib_uncompress_chksum(&self->read->decomp);
+    if (st == UZLIB_DONE) {
+        self->read->eof = true;
+    }
+    if (st < 0) {
+        DEBUG_printf("uncompress error=" INT_FMT "\n", st);
+        *errcode = MP_EINVAL;
+        return MP_STREAM_ERROR;
+    }
+    return self->read->decomp.dest - (uint8_t *)buf;
+}
+
+#if MICROPY_PY_DEFLATE_COMPRESS
+STATIC mp_uint_t deflateio_write(mp_obj_t self_in, const void *buf, mp_uint_t size, int *errcode) {
+    mp_obj_deflateio_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (self->stream == MP_OBJ_NULL || !deflateio_init_write(self)) {
```

Comment:
I swapped the ordering here of these checks, to check if stream is null first (else it would crash if a write was done after a close).